File: page_objects/TwoFASetupPage.js
import { expect } from 'chai';
import { Selector } from 'testcafe';

class TwoFASetupPage {
  static login(username, password) {
    cy.visit('/login');
    cy.get('#username').type(username);
    cy.get('#password').type(password);
    cy.get('#loginBtn').click();
  }

  static navigateToSecuritySettings() {
    cy.get('#accountMenu').click();
    cy.get('#securitySettings').click();
  }

  static verifySecuritySettingsPage() {
    cy.url().should('include', '/security-settings');
    cy.contains('Enable 2FA').should('be.visible');
  }

  static selectEnable2FA() {
    cy.get('#enable2FA').click();
  }

  static verify2FAMethodPrompt() {
    cy.contains('Choose 2FA Method').should('be.visible');
  }

  static selectSMSOption() {
    cy.get('#smsOption').click();
  }

  static verifyMobileNumberPrompt() {
    cy.contains('Enter Mobile Number').should('be.visible');
  }

  static enterMobileNumber(number) {
    cy.get('#mobileNumber').type(number);
    cy.get('#submitMobileNumber').click();
  }

  static verifyCodeSent() {
    cy.contains('Verification code sent').should('be.visible');
  }

  static enterVerificationCode(code) {
    cy.get('#verificationCode').type(code);
    cy.get('#submitVerificationCode').click();
  }

  static verifySetupCompletion() {
    cy.contains('2FA setup complete').should('be.visible');
  }

  static logout() {
    cy.get('#logoutBtn').click();
  }

  static loginOnDevice(deviceType, username, password) {
    cy.visit(`/${deviceType}-login`);
    cy.get('#username').type(username);
    cy.get('#password').type(password);
    cy.get('#loginBtn').click();
  }

  static verify2FACodePrompt() {
    cy.contains('Enter 2FA Code').should('be.visible');
  }

  static enter2FACode(code) {
    cy.get('#2FACode').type(code);
    cy.get('#submit2FACode').click();
  }

  static navigateToAccountSettings() {
    cy.get('#accountMenu').click();
    cy.get('#accountSettings').click();
  }

  static verify2FAEnabledOnAllDevices() {
    cy.contains('2FA enabled').should('be.visible');
  }

  static attemptDisable2FA() {
    cy.get('#disable2FA').click();
  }

  static verifyConfirmationCodePrompt() {
    cy.contains('Enter Confirmation Code').should('be.visible');
  }

  static enterConfirmationCode(code) {
    cy.get('#confirmationCode').type(code);
    cy.get('#submitConfirmationCode').click();
  }

  static verifyDisableConfirmation() {
    cy.contains('2FA disabled').should('be.visible');
  }

  static enable2FAUsingAuthenticatorApp() {
    cy.get('#authenticatorAppOption').click();
  }

  static verifyQRCodePrompt() {
    cy.contains('Scan QR Code').should('be.visible');
  }

  static scanQRCodeAndEnterGeneratedCode(code) {
    cy.get('#generatedCode').type(code);
    cy.get('#submitGeneratedCode').click();
  }

  static verifyAuthenticatorSetupCompletion() {
    cy.contains('Authenticator setup complete').should('be.visible');
  }

  static verifyLoginSuccessOnAllDevices() {
    cy.url().should('include', '/dashboard');
  }
}

export default TwoFASetupPage;

File: page_objects/TwoFASetupPage.java
package com.pageobjects;

import org.openqa.selenium.By;
import org.openqa.selenium.support.PageFactory;
import org.testng.Assert;
import com.framework.reusable.WebReusableComponents;

public class TwoFASetupPage extends WebReusableComponents {

    protected By btnLogin = By.id("loginButton");
    protected By dashboard = By.id("dashboard");
    protected By securitySettings = By.id("securitySettings");
    protected By enable2FAOption = By.id("enable2FA");
    protected By smsOption = By.id("smsOption");
    protected By mobileNumberField = By.id("mobileNumber");
    protected By verificationCodeField = By.id("verificationCode");
    protected By confirmationMessage = By.id("confirmationMessage");
    protected By qrCode = By.id("qrCode");
    protected By authenticatorCodeField = By.id("authenticatorCode");

    public TwoFASetupPage() {
        PageFactory.initElements(driver, this);
    }

    public void loginToApplication() {
        waitUntilElementVisible(btnLogin, 3);
        clickElement(btnLogin);
    }

    public void verifyDashboardRedirection() {
        waitUntilElementVisible(dashboard, 3);
        Assert.assertTrue(isElementVisible(dashboard), "Dashboard redirection failed.");
    }

    public void navigateToSecuritySettings() {
        waitUntilElementVisible(securitySettings, 3);
        clickElement(securitySettings);
    }

    public void verifySecuritySettingsPage() {
        waitUntilElementVisible(enable2FAOption, 3);
        Assert.assertTrue(isElementVisible(enable2FAOption), "Security settings page not displayed.");
    }

    public void selectEnable2FAOption() {
        waitUntilElementVisible(enable2FAOption, 3);
        clickElement(enable2FAOption);
    }

    public void verify2FAMethodPrompt() {
        waitUntilElementVisible(smsOption, 3);
        Assert.assertTrue(isElementVisible(smsOption), "2FA method prompt not displayed.");
    }

    public void chooseSMSMethod() {
        waitUntilElementVisible(smsOption, 3);
        clickElement(smsOption);
    }

    public void verifyMobileNumberPrompt() {
        waitUntilElementVisible(mobileNumberField, 3);
        Assert.assertTrue(isElementVisible(mobileNumberField), "Mobile number prompt not displayed.");
    }

    public void enterAndSubmitMobileNumber() {
        waitUntilElementVisible(mobileNumberField, 3);
        enterText(mobileNumberField, "1234567890");
        submitForm(mobileNumberField);
    }

    public void verifyCodeSentToMobile() {
        // Logic to verify code sent to mobile
        Assert.assertTrue(true, "Verification code sent to mobile.");
    }

    public void enterVerificationCode() {
        waitUntilElementVisible(verificationCodeField, 3);
        enterText(verificationCodeField, "123456");
        submitForm(verificationCodeField);
    }

    public void verify2FASetupCompletion() {
        waitUntilElementVisible(confirmationMessage, 3);
        Assert.assertTrue(isElementVisible(confirmationMessage), "2FA setup confirmation not displayed.");
    }

    public void verify2FASetupViaSMS() {
        // Logic to verify 2FA setup via SMS
        Assert.assertTrue(true, "2FA setup via SMS verified.");
    }

    public void logoutAndLoginOnSmartphone() {
        // Logic to logout and login on smartphone
        Assert.assertTrue(true, "Logged out and logged in on smartphone.");
    }

    public void verify2FACodePrompt() {
        waitUntilElementVisible(verificationCodeField, 3);
        Assert.assertTrue(isElementVisible(verificationCodeField), "2FA code prompt not displayed.");
    }

    public void enterCorrect2FACode() {
        waitUntilElementVisible(verificationCodeField, 3);
        enterText(verificationCodeField, "123456");
        submitForm(verificationCodeField);
    }

    public void verifySmartphoneLoginSuccess() {
        // Logic to verify smartphone login success
        Assert.assertTrue(true, "Smartphone login successful.");
    }

    public void repeatLoginAnd2FAOnTablet() {
        // Logic to repeat login and 2FA on tablet
        Assert.assertTrue(true, "Repeated login and 2FA on tablet.");
    }

    public void verifyTabletLoginSuccess() {
        // Logic to verify tablet login success
        Assert.assertTrue(true, "Tablet login successful.");
    }

    public void navigateToAccountSettings() {
        waitUntilElementVisible(securitySettings, 3);
        clickElement(securitySettings);
    }

    public void verify2FAEnabledOnAllDevices() {
        // Logic to verify 2FA enabled on all devices
        Assert.assertTrue(true, "2FA enabled on all devices.");
    }

    public void attemptToDisable2FA() {
        // Logic to attempt to disable 2FA
        Assert.assertTrue(true, "Attempted to disable 2FA.");
    }

    public void verifyConfirmationCodePrompt() {
        waitUntilElementVisible(verificationCodeField, 3);
        Assert.assertTrue(isElementVisible(verificationCodeField), "Confirmation code prompt not displayed.");
    }

    public void enterCorrectConfirmationCode() {
        waitUntilElementVisible(verificationCodeField, 3);
        enterText(verificationCodeField, "123456");
        submitForm(verificationCodeField);
    }

    public void verify2FADisabledConfirmation() {
        waitUntilElementVisible(confirmationMessage, 3);
        Assert.assertTrue(isElementVisible(confirmationMessage), "2FA disabled confirmation not displayed.");
    }

    public void verify2FADisabled() {
        // Logic to verify 2FA is disabled
        Assert.assertTrue(true, "2FA is disabled.");
    }

    public void enable2FAWithAuthenticatorApp() {
        // Logic to enable 2FA with authenticator app
        Assert.assertTrue(true, "Enabled 2FA with authenticator app.");
    }

    public void verifyQRCodePrompt() {
        waitUntilElementVisible(qrCode, 3);
        Assert.assertTrue(isElementVisible(qrCode), "QR code prompt not displayed.");
    }

    public void scanQRCodeAndEnterGeneratedCode() {
        waitUntilElementVisible(authenticatorCodeField, 3);
        enterText(authenticatorCodeField, "123456");
        submitForm(authenticatorCodeField);
    }

    public void verifyAuthenticatorAppSetupCompletion() {
        waitUntilElementVisible(confirmationMessage, 3);
        Assert.assertTrue(isElementVisible(confirmationMessage), "Authenticator app setup confirmation not displayed.");
    }

    public void verify2FASetupWithAuthenticatorApp() {
        // Logic to verify 2FA setup with authenticator app
        Assert.assertTrue(true, "2FA setup with authenticator app verified.");
    }

    public void attemptLoginOnAllDevices() {
        // Logic to attempt login on all devices
        Assert.assertTrue(true, "Attempted login on all devices.");
    }

    public void verifyLoginSuccessOnAllDevices() {
        // Logic to verify login success on all devices
        Assert.assertTrue(true, "Login successful on all devices.");
    }

    public void verify2FANotSetUp() {
        // Logic to verify 2FA is not set up
        Assert.assertTrue(true, "2FA is not set up.");
    }
}